### PR TITLE
On-click keyword and library navigation.

### PR DESCRIPTION
**Pull request by ikudriavtsev**
Supports statically created libraries and keywords. Does not support:
* Keywords, created using `@keyword` decorator
* Dynamically created keywords and libraries

### CODE REVIEW

File: robot-plugin.iml [hunk 1]
```diff
@@ -1,6 +1,18 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <module external.linked.project.id="robot-plugin" external.linked.project.path="$MODULE_DIR$" external.root.project.path="$MODULE_DIR$" external.system.id="GRADLE" external.system.module.group="" external.system.module.version="unspecified" type="JAVA_MODULE" version="4">
   <component name="DevKit.ModuleBuildProperties" url="file://$MODULE_DIR$/META-INF/plugin.xml" />
+  <component name="FacetManager">
```
Comment (AmailP) [inline robot-plugin.iml]: Ivy is not anymore used, so you can remove the facet

File: src/main/scala/amailp/intellij/robot/parser/PsiElementBuilder.scala [hunk 2]
```diff
@@ -14,7 +15,7 @@ class PsiElementBuilder(node: ASTNode) {
       case ast.SettingsTable => new psi.Settings(node)
       case ast.SettingName | ast.ResourceKey | ast.LibraryKey => new psi.SettingName(node)
       case ast.ResourceValue => new psi.ResourceValue(node)
-      case ast.LibraryValue => new psi.LibraryValue(node)
+      case ast.LibraryValue => new LibraryValue(node)
```
Comment (AmailP) [inline src/main/scala/amailp/intellij/robot/parser/PsiElementBuilder.scala]: I would leave the "psi." and remove the related import for consistency

File: src/main/scala/amailp/intellij/robot/psi/KeywordDefinition.scala [hunk 3]
```diff
@@ -45,4 +48,27 @@ object KeywordDefinition {
...
+
+  def findMatchingInLibraries(libraries: Iterable[Library], project: Project, reference: String) = findInPythonFiles(libraries, project, reference) ++
+      findInPythonClasses(libraries, project, reference)
+
+  def findInPythonFiles(libraries: Iterable[Library], project: Project, reference: String) = {
+    for {
+      library <- libraries
+      pyFile <- PyModuleNameIndex.find(library.getText, project, true)
+      keyword <- findInPythonFile(pyFile)
+      if (keyword.getName.toLowerCase matches reference.replaceAll(" ", "_").toLowerCase) && Option(keyword.getContainingClass).isEmpty
```
Comment (AmailP) [inline src/main/scala/amailp/intellij/robot/psi/KeywordDefinition.scala]: please extract a method and remove duplication for:
keyword.getName.toLowerCase matches reference.replaceAll(" ", "_").toLowerCase

File: src/main/scala/amailp/intellij/robot/psi/reference/LibraryToDefinitionReference.scala [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      .withAutoCompletionPolicy(AutoCompletionPolicy.GIVE_CHANCE_TO_OVERWRITE)
+      .asInstanceOf[AnyRef]
+  }.toArray
+
+  override def multiResolve(incompleteCode: Boolean): Array[ResolveResult] = {
+    val qName = QualifiedName.fromDottedString(library.getText)
+    var modules = mutable.MutableList[PyFile]()
+    if (qName.getComponentCount > 1) {
+      for {
+        psiFile <- PyModuleNameIndex.find(qName.getLastComponent, utilsPsiElement.getProject, true)
```
Comment (AmailP) [inline src/main/scala/amailp/intellij/robot/psi/reference/LibraryToDefinitionReference.scala]: utilsPsiElement is only used to enable the ExtRobotPsiUtils trait. Prefer here and below:
getElement.getProject

File: src/main/scala/amailp/intellij/robot/psi/reference/LibraryToDefinitionReference.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        if psiFile.getPresentation.getLocationString contains qName.getComponents.dropRight(1).mkString(".")
```
Comment (AmailP) [inline src/main/scala/amailp/intellij/robot/psi/reference/LibraryToDefinitionReference.scala]: prefer here dedicated methods, e.g:
qName.removeLastComponent().toString()
or even without toString might work

File: src/main/scala/amailp/intellij/robot/psi/reference/LibraryToDefinitionReference.scala [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package amailp.intellij.robot.psi.reference
+
+import amailp.intellij.robot.psi.utils.ExtRobotPsiUtils
+import amailp.intellij.robot.psi.LibraryValue
+import com.intellij.codeInsight.lookup.{AutoCompletionPolicy, LookupElementBuilder}
+import com.intellij.psi.{PsiElement, PsiElementResolveResult, ResolveResult, PsiPolyVariantReferenceBase}
+import com.intellij.psi.util.QualifiedName
+import com.jetbrains.python.psi.PyFile
```
Comment (AmailP) [inline src/main/scala/amailp/intellij/robot/psi/reference/LibraryToDefinitionReference.scala]: The plugin should work even without the python plugin installed (for java users for example). In this case I guess they would get a ClassDefNotFound exception that is not what should happen. If you look at the moment the com.jetbrains.python.* classes are only loaded when the optional python plugin is present via an optional plugin desctriptor (META-INF/robot-python.xml)

File: src/main/scala/amailp/intellij/robot/psi/KeywordDefinition.scala [hunk 7]
```diff
@@ -2,8 +2,11 @@ package amailp.intellij.robot.psi
 
 import amailp.intellij.robot.ast
 import com.intellij.lang.ASTNode
+import com.intellij.openapi.project.Project
 import com.intellij.psi._
 import com.intellij.psi.util.PsiTreeUtil
+import com.jetbrains.python.psi.{PyFunction, PyFile}
```
Comment (AmailP) [inline src/main/scala/amailp/intellij/robot/psi/KeywordDefinition.scala]: The plugin should work even without the python plugin installed (for java users for example). In this case I guess they would get a ClassDefNotFound exception that is not what should happen. If you look at the moment the com.jetbrains.python.* classes are only loaded when the optional python plugin is present via an optional plugin desctriptor (META-INF/robot-python.xml)

File: src/main/scala/amailp/intellij/robot/psi/reference/LibraryToDefinitionReference.scala [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    } yield LookupElementBuilder.create(lib.getText)
+      .withCaseSensitivity(false)
+      .withTypeText("Library", true)
+      .withAutoCompletionPolicy(AutoCompletionPolicy.GIVE_CHANCE_TO_OVERWRITE)
+      .asInstanceOf[AnyRef]
+  }.toArray
+
+  override def multiResolve(incompleteCode: Boolean): Array[ResolveResult] = {
+    val qName = QualifiedName.fromDottedString(library.getText)
+    var modules = mutable.MutableList[PyFile]()
```
Comment (AmailP) [inline src/main/scala/amailp/intellij/robot/psi/reference/LibraryToDefinitionReference.scala]: you do not need a mutable list here. looks like you can just use the return value of the **if** expression:

```scala
var module = if(...) {
  for {
    ...
  } *yield* psiFile
} else {
  PyModuleNameIndex.find(...)
}
```

**Comment by AmailP** [comment]
This resolves #10

File: src/main/scala/amailp/intellij/robot/psi/reference/LibraryToDefinitionReference.scala [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+import com.jetbrains.python.psi.stubs.{PyClassNameIndex, PyModuleNameIndex}
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+import amailp.intellij.robot.psi.utils.ExtRobotPsiUtils
+
+
+class LibraryToDefinitionReference(element: PsiElement, textRange: TextRange) extends PsiReferenceBase[PsiElement](element, textRange)
+  with PsiPolyVariantReference with ExtRobotPsiUtils {
+
+  override def getVariants = {
```
Comment (AmailP) [inline src/main/scala/amailp/intellij/robot/psi/reference/LibraryToDefinitionReference.scala]: the getVariants serves the purpose of populate the autocompletion list. What is the point in populating it with the list of the already imported libraries?
Theoretically all the reachable python modules and classes are candidates for completion but the ones already imported are the less interesting, don't you think?
I would rather return an empty result...

File: src/main/resources/META-INF/plugin.xml [hunk 10]
```diff
@@ -169,6 +169,9 @@
 
         <!-- Structure view -->
         <lang.psiStructureViewFactory language="Robotframework" implementationClass="amailp.intellij.robot.extensions.RobotPsiStructureViewFactory"/>
+
+        <!-- Robot keyword reference -->
+        <psi.referenceContributor implementation="amailp.intellij.robot.psi.reference.KeywordReferenceContributor"/>
```
Comment (AmailP) [inline src/main/resources/META-INF/plugin.xml]: Why is this needed? AFAIK the reference contributor serves for non standard references but references are usually retrieved through the getReference method as they currently are. See [here](https://github.com/JetBrains/intellij-community/blob/master/platform/core-api/src/com/intellij/psi/PsiReference.java#L27)

Comment (ikudriavtsev) [inline src/main/resources/META-INF/plugin.xml]: This is needed to make the Python API usage pluggable. There is another reference contributor registered in src/main/resources/META-INF/robot-python.xml. If I leave `getReference` defined in the class then the python reference results will be overridden.

File: src/main/scala/amailp/intellij/robot/psi/reference/LibraryToDefinitionReference.scala (hunk 9 above)
Comment (ikudriavtsev) [inline src/main/scala/amailp/intellij/robot/psi/reference/LibraryToDefinitionReference.scala]: Well spotted! I think returning all reachable modules and classes is a bit too much, so I'd go for an empty list.

File: src/main/resources/META-INF/plugin.xml (hunk 10 above)
Comment (AmailP) [inline src/main/resources/META-INF/plugin.xml]: Ok. I am merging your PR locally on my machine, adding a bunch of tests and refactoring a bit. I will then (in a few days) publish a merged branch for you to review, ok?

Comment (ikudriavtsev) [inline src/main/resources/META-INF/plugin.xml]: Sure, thank you.

**Comment by ikudriavtsev** [comment]
Hi,

any update on this one?

Thanks.

**Comment by AmailP** [comment]
Yes... sorry for the LONG delay :-( but the little one is sucking every
second of my free time lately...

I had added some months ago some tests to your implementation and
refactored it in origin/feat_jump_to_python. I have found the time to
upload it just now :-/

The test are currently only for the jump to library. It would be nice to
add also tests for the jump to python keywords. They should be not too
different from the former ones.

Please feel free to comment on my rework and extend it.

Valerio

**Comment by AmailP** [comment]
The PR has been refactored and merged in master
